cpp/dolfinx/fem/discreteoperators.h

```diff
@@ -68,128 +66,78 @@ void assemble_discrete_gradient(
-  auto e_to_v = mesh->topology().connectivity(1, 0);
-  mesh->topology_mutable().create_connectivity(tdim, 1);
-  auto c_to_e = mesh->topology().connectivity(tdim, 1);
-  mesh->topology_mutable().create_connectivity(1, tdim);
-  auto e_to_c = mesh->topology().connectivity(1, tdim);
-  mesh->topology_mutable().create_connectivity(tdim, 0);
-  auto c_to_v = mesh->topology().connectivity(tdim, 0);
-
-  // Build sparsity pattern
-  const std::int32_t num_edges = mesh->topology().index_map(1)->size_local()
-                                 + mesh->topology().index_map(1)->num_ghosts();
-  const std::shared_ptr<const fem::DofMap> dofmap0 = V0.dofmap();
-  assert(dofmap0);
-  // Create local lookup table for local edge to cell dofs
-  const int num_edges_per_cell
-      = mesh::cell_num_entities(mesh->topology().cell_type(), 1);
-  std::map<std::int32_t, std::vector<std::int32_t>> local_edge_dofs;
-  for (std::int32_t i = 0; i < num_edges_per_cell; ++i)
-    local_edge_dofs[i] = layout0.entity_dofs(1, i);
-  // Create local lookup table for local vertex to cell dofs
-  const int num_vertices_per_cell
-      = mesh::cell_num_entities(mesh->topology().cell_type(), 0);
-  std::map<std::int32_t, std::vector<std::int32_t>> local_vertex_dofs;
-  for (std::int32_t i = 0; i < num_vertices_per_cell; ++i)
-    local_vertex_dofs[i] = layout1.entity_dofs(0, i);
-
-  // Build discrete gradient operator/matrix
-  const std::shared_ptr<const fem::DofMap> dofmap1 = V1.dofmap();
-  assert(dofmap1);
-  const std::vector<std::int64_t>& global_indices
-      = mesh->topology().index_map(0)->global_indices();
-  std::array<T, 2> Ae;
-  for (std::int32_t e = 0; e < num_edges; ++e)
+  std::array<std::size_t, 4> t_shape
+      = {(std::size_t)tdim + 1, X.shape(0), (std::size_t)ndofs_cell_1,
+         (std::size_t)ref_vs_1};
+  xt::xtensor<double, 4> l_basis(t_shape);
+  e1->tabulate(l_basis, X, 1);
+  auto l_dphi
+      = xt::view(l_basis, xt::xrange(1, tdim + 1), xt::all(), xt::all(), 0);
+  auto dphi = xt::reshape_view(l_dphi, {tdim * t_shape[1], t_shape[2]});
+
+  // Compute local element interpolation matrix
+  const int bs_0 = e0->block_size();
+  assert(bs_0 == 1);
+  const int ndofs_cell_0 = e0->space_dimension() / bs_0;
+  xt::xtensor<T, 2> Ae = xt::zeros<double>({ndofs_cell_0, ndofs_cell_1});
+  const xt::xtensor<double, 2> Pi = e0->interpolation_operator();
+  math::dot(Pi, dphi, Ae);
+
+  // Get inverse transform
+  const std::function<void(const xtl::span<T>&,
+                           const xtl::span<const std::uint32_t>&, std::int32_t,
+                           int)>
+      apply_inverse_dof_transform
+      = e0->get_dof_transformation_function<T>(true, true, false);
+  mesh->topology_mutable().create_entity_permutations();
+  xtl::span<const std::uint32_t> cell_info
+      = xtl::span(mesh->topology().get_cell_permutation_info());
+  auto dofmap0 = V0.dofmap();
+  auto dofmap1 = V1.dofmap();
+  // Create element kernel
```

Comment:
This comment is a bit vague.

In general, comments should have a black line before them for readability.